server/src/api/bill_data.py

```diff
@@ -60,8 +81,104 @@ def get_bills(self):
-            error_msg = "User not logged in"
+            #Check if list should be ascending or descending
+            if(sort_order == "asc"):
+                descending=False
+            elif(sort_order == "desc"):
+                descending=True
+            else:
+                #Invalid API call
+                raise InvalidDataException("sort_order")
+
+            #Sort list by listed sort_field
+            sort=[]
+            if(sort_field in bill_sort_fields):
+                #These fields only require searching the bills dict
+
+                #Translate the sort_field to the related dict key
+                if(sort_field=="amount"):
+                    field="total"
+                elif(sort_field=="created time"):
+                    field="timestamp"
+                elif (sort_field=="name"):
+                    field="title"
+                    #Due to how ascii would end up at the front of an a descending sort
+                    #to make more intuitive sense the order is flipped for names
+                    descending=not(descending)
+
+                for i in bills:
+                    if (len(sort) == 0 ):
+                            sort.append(i)
+                            continue
+                    for j in range( 0,len(sort)+1 ):
+
+                        if(j == len(sort)):
+                            #reached the end of the list, add it to the end
+                            sort.append(i)
+                        else:
+                            #When comparing strings we would like to ignore capitilisation so that
+                            #the ascii values dont sort A,a,B to A,B,a
+                            #however lower only exists on strings, otherwise it will throw an error
+                            if(sort_field=="name"):
+                                if( ( (i[field]).lower() > (sort[j][field]).lower() ) == descending):
+                                    sort.insert(j,i)
+                                    break
+                            else:
+                                if( ( (i[field]) > (sort[j][field]) ) == descending):
+                                    sort.insert(j,i)
+                                    break
+            else:
+                #Invalid API call
+                raise InvalidDataException("sort_field")
+
+            bills=sort
+
+            #check for names in the payee and payer fields
+            filter_payee=( len(payee) > 0 )
+            filter_payer=( len(payer) > 0 )
+            #Remove paid/unpaid bills if requested
+            output=[]
+            if(is_paid == "paid" or is_paid=="unpaid" or is_paid=="either"):
+                for i in bills:
+                    valid=True
+                    valid_payee=False
+                    valid_payer=False
+                    for j in i["payments"]:
+                        if( not( j["is_paid"] == (is_paid == "paid") ) and is_paid != "either"):
```

Comment:
I'd suggest some comments to make this line easier to read. It was hard for me to understand what it was doing upon the first read. It was also confusing since the 'valid' flag is switched differently to the proceeding "if" statements.